### DISCUSSION TITLE
`numeric` literal possible?

### DISCUSSION THREAD

**Discussion by boredland** [elysiajs/elysia#915] [Q&A]
Hi there! I would like to ensure that a parameter is part of a list of numeric values.

I tried something along the lines of:

```typescript
const NumberEnum = <T extends number[]>(values: [...T]) =>
  t.Unsafe<T[number]>(t.Union(values.map((value) => t.Literal(value))));
```

Which of course only allows numbers _in_, but produces the correct type at least.

Has anybody done that?

**Comment by crishoj** [comment]
Have you tried...

```ts
const NumberEnum = <T extends number[]>(values: [...T]) => t.Unsafe<T[number]>({
    type: 'number', enum: values
})
```

<img width="520" alt="image" src="https://github.com/user-attachments/assets/a0f4e9f5-a1c0-4592-beb4-34de65f2be95">

**Reply by boredland** [reply]
Yes, sadly elysia seems to not allow me to define any schemas myself:

![image](https://github.com/user-attachments/assets/da899966-8448-44cb-a9b6-d46f006b4fe3)

**Reply by crishoj** [reply]
Could you post a minimal code example to reproduce the error?

**Reply by boredland** [reply]
Sure, hope a basic bun test is sufficient (this is the string-enum example from the typebox docs):

```
import { expect, test } from 'bun:test';
import { Value } from '@sinclair/typebox/value';
import { t } from 'elysia';

test('validate string enum', () => {
  const StringEnum = <T extends string[]>(values: [...T]) =>
    t.Unsafe<T[number]>({
      type: 'string',
      enum: values,
    });
  const S = StringEnum(['A', 'B', 'C']);

  expect(Value.Check(S, 'A')).toBe(true);
  expect(Value.Check(S, 'B')).toBe(true);
  expect(Value.Check(S, 'C')).toBe(true);
  expect(Value.Check(S, 'D')).toBe(false);
});
```

**Reply by crishoj** [reply]
For runtime checks, a Typebox union of literals does the trick:

```ts
    const allowed = ['A', 'B', 'C']
    const S = t.Union(allowed.map((v) => t.Literal(v)))
```

**Reply by boredland** [reply]
so `t.Unsafe` isn't working?

**Reply by crishoj** [reply]
My understanding of `t.Unsafe` is lacking.

As far as I can tell, `Value.Check` requires the schema to have a `[Kind]` property matching an entry in Typebox' type registry:

https://github.com/sinclairzx81/typebox/blob/4e2158deb238b1fe958b0f78ad9e05f21e45041c/src/value/check/check.ts#L484

**Reply by boredland** [reply]
So as I get it, a union of "numerics", parsed as a number, isn't possible currently.

**Reply by crishoj** [reply]
I think you're right.

Assuming I understand your requirement correctly, this was what I came up with:

```ts
import Elysia, {t} from 'elysia'

const allowed = [1, 2, 3]
const schema = t.Object({
    num: t.Union(allowed.map(n => t.Literal(`${n}`)))
})

const app = new Elysia()
    .resolve(({params}) => params.num = Number.parseInt(params.num))
    .get('/test/:num', ({params: {num}}) => `num = ${num} (${typeof num})`, {
        params: schema
    })

const urls = [
    '/test/1',
    '/test/2',
    '/test/3',
    '/test/9',
]

for (const url of urls) {
    const res = await app.handle(new Request(`https://localhost${url}`))
    if (res.ok) {
        console.log(`GET ${url} → HTTP ${res.status} ${await res.text()}`)
    } else {
        console.log(`GET ${url} → HTTP ${res.status} ${(await res.json()).summary}`)
    }
}
```

```
GET /test/1 → HTTP 200 num = 1 (number)
GET /test/2 → HTTP 200 num = 2 (number)
GET /test/3 → HTTP 200 num = 3 (number)
GET /test/9 → HTTP 422 Property 'num' should be one of: 'string', 'string', 'string'
```

Curious to see a more idiomatic approach.
